File: Backtracking/path_with_maximum_gold.cpp
// https://leetcode.com/problems/path-with-maximum-gold/

// Solution 1
class Solution {
public:
    int dx[4] = { 1, -1, 0, 0};
    int dy[4] = { 0, 0, 1, -1};
    
    bool OK(vector<vector<int>>& grid, int i, int j) {
        if(i < 0 || i >= grid.size() || j < 0 || j >= grid[i].size())
            return false;
        
        if(!grid[i][j])
            return false;
        
        return true;
    }
    
    int solve(vector<vector<int>>& grid, int x, int y) {        
        int aux = grid[x][y];
        grid[x][y] = 0;
        int res = 0;
        for(int i = 0; i < 4; i++) {
            int nextX = x + dx[i];
            int nextY = y + dy[i];
            if(OK(grid, nextX, nextY)) {
                res = max(res, grid[nextX][nextY] + solve(grid, nextX, nextY));
            }
        }
        grid[x][y] = aux;
        return res;
    }
    
    int getMaximumGold(vector<vector<int>>& grid) {
        int res = 0;
        
        for(int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[i].size(); j++) {
                if(grid[i][j])
                    res = max(res, grid[i][j] + solve(grid, i, j));
            }
        }
        
        return res;
    }
};

// Solution 2
class Solution {
public:
    int dx[4] = { 1, -1, 0, 0};
    int dy[4] = { 0, 0, 1, -1};
    int res = 0;
    
    bool OK(vector<vector<int>>& grid, int i, int j) {
        if(i < 0 || i >= grid.size() || j < 0 || j >= grid[i].size())
            return false;
        
        if(!grid[i][j])
            return false;
        
        return true;
    }
    
    void solve(vector<vector<int>>& grid, int x, int y, int sum) {
        res = max(res, sum);
        
        int aux = grid[x][y];
        grid[x][y] = 0;
        for(int i = 0; i < 4; i++) {
            int nextX = x + dx[i];
            int nextY = y + dy[i];
            if(OK(grid, nextX, nextY)) {
                solve(grid, nextX, nextY, sum + grid[nextX][nextY]);
            }
        }
        grid[x][y] = aux;
    }
    
    int getMaximumGold(vector<vector<int>>& grid) {
        for(int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[i].size(); j++) {
                if(grid[i][j])
                    solve(grid, i, j, grid[i][j]);
            }
        }
        
        return res;
    }
};

File: Graph/path_with_maximum_gold.go
// https://leetcode.com/problems/path-with-maximum-gold

var (
    dx = []int{1, -1, 0, 0}
    dy = []int{0, 0, 1, -1}
)

func canMove(grid [][]int, x, y int) bool {
    if x < 0 || x >= len(grid) {
        return false
    }

    if y < 0 || y >= len(grid[x]) {
        return false
    }

    return grid[x][y] > 0
}

func solve(grid [][]int, x, y int) int {
    gold := grid[x][y]
    grid[x][y] = 0

    ans := gold
    for k := 0; k < len(dx); k++ {
        i := x + dx[k]
        j := y + dy[k]
        if canMove(grid, i, j) {
            ans = max(ans, gold + solve(grid, i, j))
        }
    }

    grid[x][y] = gold

    return ans
}

func getMaximumGold(grid [][]int) int {
    ans := 0

    for i := 0; i < len(grid); i++ {
        for j := 0; j < len(grid[i]); j++ {
            if grid[i][j] > 0 {
                ans = max(ans, solve(grid, i, j))
            }
        }
    }

    return ans
}
